content/blog/2020-04-dplyr-1-0-0-and-vctrs/index.Rmarkdown

```diff
@@ -0,0 +1,197 @@
+
+
+Today, I wanted to talk a little bit how dplyr 1.0.0 uses the  [vctrs](http://vctrs.r-lib.org/) package. This post explains why vctrs is so important, why we can't just copy what base R does, how to interpret some of new error messages that you'll see, and some of the major changes since the last version.
+
+If you're interested in living life on the edge (or trying out anything you see in this blog post), you can install the development version of dplyr with:
+
+```{r, eval = FALSE}
+devtools::install_github("tidyverse/dplyr")
+```
+
+Note that the development version won't become 1.0.0 until it's released, but it has all the same features.
+
+```{r setup}
+library(vctrs)
+library(dplyr, warn.conflicts = FALSE)
+```
+
+## Combining vectors
+
+The heart of the reason we're using vctrs is the need to combine vectors together. You're already familiar with one base R tool for combining vectors, `c()`:
+
+```{r}
+c(1, 2, 3)
+c("a", "b", "c")
+```
+
+But combining vectors comes up in many places in the tidyverse, e.g.:
+
+* `dplyr::mutate()` and `dplyr::summarise()` have to combine the results 
+  from each group.
+* `dplyr::bind_rows()` has to combine columns from different data frames.
+* `dplyr::full_join()`  has to combine the keys from the `x` and `y` data
+  frames.
+* `tidyr::pivot_longer()` has to combine multiple columns into one.
+
+Our goal is to unify the code that underlies all these various functions so that there's one consistent, principled approach. We've already made the change in [tidyr](https://www.tidyverse.org/blog/2019/09/tidyr-1-0-0/), and now it's dplyr's turn. 
+
+## Base R behaviour
+
+You might wonder why we can't just copy the behaviour of `c()`. Unfortunately `c()` has some major downsides:
+
+*   It doesn't possess a `factor` method so it converts factors to their
+    underlying integer levels. 
+    
+    ```{r}
+    c(factor("x"), factor("y"))
+    ```
+
+*   It's difficult to implement methods when different classes are involved.
+    For example, combining a date (`Date`) and a date-time (`POSIXct`) yields
+    an incorrect result because the underlying data is combined without
+    first being translated.
+
+    ```{r}
+    today <- as.Date("2020-03-24")
+    now <- as.POSIXct("2020-03-24 10:34")
+    
+    c(today, now)
+    # (the second value is the 11 Dec 4341727-12-11)
+    class(c(today, now))
+    unclass(c(today, now))
+    
+    c(now, today)
+    class(c(now, today))
+    unclass(c(now, today))
+    ```
+
+It's difficult to change how `c()` works because any changes are likely to break some existing code, and base R is fundamentally conservative. Additionally, `c()` isn't the only way that base R combines vectors. `rbind()` and `unlist()` can also be used to perform a similar job, but return different results. This is not to say that the tidyverse has been any better in the past --- we have used a variety of ad hoc methods, undoubtedly using well more than three different approaches. 
```

Comment:
```suggestion
It's difficult to change how `c()` works because any changes are likely to break some existing code, and base R is committed to backward compatibility. Additionally, `c()` isn't the only way that base R combines vectors. `rbind()` and `unlist()` can also be used to perform a similar job, but return different results. This is not to say that the tidyverse has been any better in the past --- we have used a variety of ad hoc methods, undoubtedly using well more than three different approaches. 
```

More positive phrasing.